File: Pandacap.Data/MyContent/AddressedPost.fs
namespace Pandacap.Data

open System
open System.ComponentModel.DataAnnotations.Schema
open Pandacap.Html

type AddressedPost() =
    member val Id = Guid.Empty with get, set
    member val InReplyTo = nullString with get, set
    member val Community = nullString with get, set
    member val Users = new ResizeArray<string>() with get, set
    member val PublishedTime = DateTimeOffset.MinValue with get, set
    member val Title = nullString with get, set
    member val HtmlContent = "" with get, set
    member val IsDirectMessage = false with get, set

    [<NotMapped>]
    member this.IsReply = not (isNull this.InReplyTo)

    [<NotMapped>]
    member this.Addressing =
        let users = this.Users |> Seq.toList
        let communities = this.Community |> Option.ofObj |> Option.toList

        if this.IsDirectMessage then
            {|
                To = users @ communities
                Cc = []
            |}
        else if this.IsReply then
            {|
                To = "https://www.w3.org/ns/activitystreams#Public" :: []
                Cc = users @ communities
            |}
        else
            {|
                To = "https://www.w3.org/ns/activitystreams#Public" :: communities
                Cc = []
            |}

    interface IPost with
        member _.Badges = []
        member this.DisplayTitle = ExcerptGenerator.FromText 120 (seq {
            this.Title
            TextConverter.FromHtml this.HtmlContent
            $"{this.Id}"
        })
        member this.Id = $"{this.Id}"
        member this.LinkUrl = $"/AddressedPosts/{this.Id}"
        member _.ProfileUrl = null
        member _.Thumbnails = []
        member this.Timestamp = this.PublishedTime
        member _.Usericon = null
        member _.Username = null

    interface Pandacap.ActivityPub.IAddressedPost with
        member this.Audience = this.Community
        member this.Cc = this.Addressing.Cc
        member this.Html = this.HtmlContent
        member this.Id = this.Id
        member this.InReplyTo = this.InReplyTo
        member this.PublishedTime = this.PublishedTime
        member this.Title = this.Title
        member this.To = this.Addressing.To

File: Pandacap.Database/AddressedPost.cs
using Microsoft.FSharp.Collections;
using Pandacap.ActivityPub.Models.Interfaces;
using Pandacap.ActivityPub.Static;
using Pandacap.Text;
using Pandacap.UI.Badges;
using Pandacap.UI.Elements;
using System.ComponentModel.DataAnnotations.Schema;

namespace Pandacap.Database
{
    public class AddressedPost : IPost, IActivityPubPost, IActivityPubAddressing
    {
        public Guid Id { get; set; }

        public string? InReplyTo { get; set; }

        public string? Community { get; set; }

        public List<string> Users { get; set; } = [];

        public DateTimeOffset PublishedTime { get; set; }

        public string? Title { get; set; }

        public string HtmlContent { get; set; } = "";

        public bool IsDirectMessage { get; set; }

        public string? BlueskyDID { get; set; }

        public string? BlueskyRecordKey { get; set; }

        [NotMapped]
        public bool IsReply => InReplyTo != null;

        private IEnumerable<string> EnumerateRecipients()
        {
            foreach (var user in Users)
                yield return user;

            if (Community is string c)
                yield return c;
        }

        public record PostAddressing(
            FSharpList<string> To,
            FSharpList<string> Cc);

        [NotMapped]
        public PostAddressing Addressing =>
            IsDirectMessage
                ? new(
                    To: [.. EnumerateRecipients()],
                    Cc: [])
            : IsReply
                ? new(
                    To: ["https://www.w3.org/ns/activitystreams#Public"],
                    Cc: [.. EnumerateRecipients()])
                : new(
                    To: ["https://www.w3.org/ns/activitystreams#Public"],
                    Cc: []);

        Badge IPost.Badge => Badges.ActivityPub;

        string IPost.DisplayTitle => !string.IsNullOrWhiteSpace(Title)
            ? Title
            : ExcerptGenerator.FromText(60, TextConverter.FromHtml(HtmlContent));

        string IPost.Id => $"{Id}";

        string IPost.InternalUrl => $"/AddressedPosts/{this.Id}";

        string IPost.ExternalUrl => $"/AddressedPosts/{this.Id}";

        DateTimeOffset IPost.PostedAt => PublishedTime;

        string? IPost.ProfileUrl => null;

        IEnumerable<IPostThumbnail> IPost.Thumbnails => [];

        string? IPost.Username => null;

        string? IPost.Usericon => null;

        string IActivityPubPost.ObjectId => $"https://{ActivityPubHostInformation.ApplicationHostname}/AddressedPosts/{Id}";

        IActivityPubAddressing IActivityPubPost.Addressing => this;

        bool IActivityPubPost.IsArticle => false;

        string IActivityPubPost.Html => HtmlContent;

        IEnumerable<string> IActivityPubPost.Tags => [];

        IEnumerable<IActivityPubLink> IActivityPubPost.Links => [];

        IEnumerable<IActivityPubImage> IActivityPubPost.Images => [];

        IEnumerable<string> IActivityPubAddressing.To => Addressing.To;

        IEnumerable<string> IActivityPubAddressing.Cc => Addressing.Cc;

        string? IActivityPubAddressing.Audience => Community;
    }
}
